### PR TITLE
[IOTDB-4448] Implement pattern based data deletion in DataRegion to speed up deleting timeseries

### PR DESCRIPTION
**Pull request by MarcosZyk**
### Modification
Implement pattern based data deletion in DataRegion to speed up deleting timeseries. When encountering path pattern, deviceIds in memtable will be traversed and those matched by given path pattern will be chosen for further data deletion execution.

Clean up all datanode schema cache, which much faster than clean precise timeseries.

### CODE REVIEW

File: server/src/main/java/org/apache/iotdb/db/engine/storagegroup/DataRegion.java [hunk 1]
```diff
@@ -2332,6 +2332,9 @@ private boolean canSkipDelete(
     }
 
     for (PartialPath device : devicePaths) {
+      if (device.hasWildcard()) {
```
Comment (qiaojialin) [inline server/src/main/java/org/apache/iotdb/db/engine/storagegroup/DataRegion.java:2335]: check if the time cost of precise matching is tolerable.

**Review by qiaojialin — CHANGES_REQUESTED** [review]

File: server/src/main/java/org/apache/iotdb/db/engine/storagegroup/DataRegion.java (hunk 1 above)
Comment (MarcosZyk) [reply server/src/main/java/org/apache/iotdb/db/engine/storagegroup/DataRegion.java:2336]: Basically, the performance of query has higher priority than that of data deletion. From the experiment result, it costs 280ms to execute precise matching in memory for 10000 devices，each having 1 measurement and 1000 points.

File: confignode/src/main/java/org/apache/iotdb/confignode/procedure/impl/DeleteTimeSeriesProcedure.java [hunk 2]
```diff
@@ -94,38 +94,46 @@ public DeleteTimeSeriesProcedure(String queryId, PathPatternTree patternTree) {
...
+          deleteData(env);
           break;
-        } else {
-          setFailure(
-              new ProcedureException(
-                  new PathNotExistException(
-                      patternTree.getAllPathPatterns().stream()
-                          .map(PartialPath::getFullPath)
-                          .collect(Collectors.toList()))));
+        case DELETE_TIMESERIES:
```
[the remark this answers is not in the document]
Comment (MarcosZyk) [reply confignode/src/main/java/org/apache/iotdb/confignode/procedure/impl/DeleteTimeSeriesProcedure.java]: Renamed to DELETE_TIMESERIES_SCHEMA.

File: confignode/src/main/java/org/apache/iotdb/confignode/procedure/impl/DeleteTimeSeriesProcedure.java [hunk 3]
```diff
@@ -94,38 +94,46 @@ public DeleteTimeSeriesProcedure(String queryId, PathPatternTree patternTree) {
...
+          LOGGER.info("Delete timeseries schema of {}", requestMessage);
+          deleteTimeSeries(env);
```
[the remark this answers is not in the document]
Comment (MarcosZyk) [reply confignode/src/main/java/org/apache/iotdb/confignode/procedure/impl/DeleteTimeSeriesProcedure.java]: Fixed.

File: server/src/main/java/org/apache/iotdb/db/engine/memtable/AbstractMemTable.java [hunk 4]
```diff
@@ -658,13 +642,43 @@ public ReadOnlyMemChunk query(
   @Override
   public void delete(
       PartialPath originalPath, PartialPath devicePath, long startTimestamp, long endTimestamp) {
```
[the remark this answers is not in the document]
Comment (MarcosZyk) [reply server/src/main/java/org/apache/iotdb/db/engine/memtable/AbstractMemTable.java:652]: Added.

File: server/src/main/java/org/apache/iotdb/db/engine/memtable/AbstractMemTable.java [hunk 5]
```diff
@@ -654,17 +638,53 @@ public ReadOnlyMemChunk query(
         .getReadOnlyMemChunkFromMemTable(this, modsToMemtable, ttlLowerBound);
   }
 
+  /**
+   * @param originalPath the original path pattern or full path to be used to match timeseries
```
[the remark this answers is not in the document]
Comment (MarcosZyk) [reply server/src/main/java/org/apache/iotdb/db/engine/memtable/AbstractMemTable.java]: Fixed.

File: server/src/main/java/org/apache/iotdb/db/engine/memtable/AbstractMemTable.java [hunk 6]
```diff
@@ -654,17 +638,53 @@ public ReadOnlyMemChunk query(
...
+   * @param devicePath the device path of the timeseries to be deleted.
```
[the remark this answers is not in the document]
Comment (MarcosZyk) [reply server/src/main/java/org/apache/iotdb/db/engine/memtable/AbstractMemTable.java]: Fixed.

File: server/src/main/java/org/apache/iotdb/db/metadata/mtree/MTreeBelowSGMemoryImpl.java [hunk 7]
```diff
@@ -468,6 +468,22 @@ protected void collectMeasurement(IMeasurementMNode node) throws MetadataExcepti
     return result;
   }
 
+  public Set<PartialPath> getDevicesOfPreDeletedTimeseries(PartialPath pathPattern)
```
[the remark this answers is not in the document]
Comment (MarcosZyk) [reply server/src/main/java/org/apache/iotdb/db/metadata/mtree/MTreeBelowSGMemoryImpl.java:476]: Doc added.

**Review by qiaojialin — APPROVED** [review]

**Review by qiaojialin — APPROVED** [review]

**Review by qiaojialin — APPROVED** [review]

**Pull request merged**